protocol/tx.go

```diff
@@ -1,61 +1,40 @@
 package protocol
 
 import (
-	"sync"
-
-	"github.com/golang/groupcache/lru"
-
 	"chain/errors"
 	"chain/protocol/bc"
-	"chain/protocol/validation"
 )
 
-// ValidateTxCached checks a cache of prevalidated transactions
-// before attempting to perform a context-free validation of the tx.
-func (c *Chain) ValidateTxCached(tx *bc.Tx) error {
-	// Consult a cache of prevalidated transactions.
-	err, ok := c.prevalidated.lookup(tx.ID)
-	if ok {
-		return err
-	}
+// ErrBadTx is returned for transactions failing validation
+var ErrBadTx = errors.New("invalid transaction")
 
-	err = validation.CheckTxWellFormed(tx)
-	c.prevalidated.cache(tx.ID, err)
-	return err
-}
-
-type prevalidatedTxsCache struct {
-	mu  sync.Mutex
-	lru *lru.Cache
-}
-
-func (c *prevalidatedTxsCache) lookup(txID bc.Hash) (err error, ok bool) {
-	c.mu.Lock()
-	v, ok := c.lru.Get(txID)
-	c.mu.Unlock()
-	if !ok {
-		return err, ok
-	}
-	if v == nil {
-		return nil, ok
+func (c *Chain) checkIssuanceWindow(tx *bc.TxEntries) error {
+	if c.MaxIssuanceWindow == 0 {
+		return nil
 	}
-	return v.(error), ok
-}
-
-func (c *prevalidatedTxsCache) cache(txID bc.Hash, err error) {
-	c.mu.Lock()
-	c.lru.Add(txID, err)
-	c.mu.Unlock()
-}
-
-func (c *Chain) checkIssuanceWindow(tx *bc.Tx) error {
-	for _, txi := range tx.Inputs {
-		if _, ok := txi.TypedInput.(*bc.IssuanceInput); ok {
-			// TODO(tessr): consider removing 0 check once we can configure this
-			if c.MaxIssuanceWindow != 0 && tx.MinTime+bc.DurationMillis(c.MaxIssuanceWindow) < tx.MaxTime {
-				return errors.WithDetailf(validation.ErrBadTx, "issuance input's time window is larger than the network maximum (%s)", c.MaxIssuanceWindow)
+	for _, entry := range tx.TxInputs {
+		if _, ok := entry.(*bc.Issuance); ok {
+			if tx.Body.MinTimeMS+bc.DurationMillis(c.MaxIssuanceWindow) < tx.Body.MaxTimeMS {
+				return errors.WithDetailf(ErrBadTx, "issuance input's time window is larger than the network maximum (%s)", c.MaxIssuanceWindow)
 			}
 		}
 	}
 	return nil
 }
+
+// ValidateTx validates the given transaction. A cache holds
+// per-transaction validation results and is consulted before
+// performing full validation.
+func (c *Chain) ValidateTx(tx *bc.TxEntries) error {
+	err := c.checkIssuanceWindow(tx)
+	if err != nil {
+		return err
+	}
+	var ok bool
+	err, ok = c.prevalidated.lookup(tx.ID)
+	if !ok {
+		err = bc.ValidateTx(tx, c.InitialBlockHash)
```

Comment:
Not without pain.

As you know, this PR started out much bigger, but I factored out a series of smaller lead-up PR's, each a logical stepping stone in getting to this one. Changing the validation API, however, did not happen in isolation from switching to entries. Instead, I wrote a whole new validate/apply API, trying simultaneously to make it line up with existing callers' expectations and to clarify (compared to what we had before) the boundaries and contracts of the different entrypoints.

In other words, restoring the previous validation API would be brand-new work, unlike all the lead-up PRs. It's not impossible, and I'll do it if you really think it's needed, but since the intent would be quickly to replace it with what's already in this PR, my strong preference is not to spend the effort.